From 2f95510a8dc0bb1e97a4dbaf23cd5e0d756b4542 Mon Sep 17 00:00:00 2001
From: Hong-Mei Li <a21834@motorola.com>
Date: Thu, 28 May 2015 03:33:30 -0700
Subject: [PATCH 190/959] IKSWM-4458 msm-poweroff: handle special reboot cmds

For post-panic, post-wdt, post-pmicwdt reboot, we do cold reset.
And set the PMIC register accordingly, to nofity BL the special
reboot behavior.

Change-Id: Ib40e3e10efb783054bf79d9fb3e2c99de608f6ef
Signed-off-by: Hong-Mei Li <a21834@motorola.com>
Reviewed-on: http://gerrit.mot.com/752079
SME-Granted: SME Approvals Granted
SLTApproved: Slta Waiver <sltawvr@motorola.com>
Tested-by: Jira Key <jirakey@motorola.com>
Reviewed-by: Zhi-Ming Yuan <a14194@motorola.com>
Reviewed-by: Dong-Hua Yan <a18317@motorola.com>
Submit-Approved: Jira Key <jirakey@motorola.com>
---
 drivers/power/reset/msm-poweroff.c | 18 ++++++++++++++++++
 include/linux/qpnp/power-on.h      |  4 ++++
 2 files changed, 22 insertions(+)

diff --git a/drivers/power/reset/msm-poweroff.c b/drivers/power/reset/msm-poweroff.c
index 447a05f80e3..6fcf0c5ed0b 100644
--- a/drivers/power/reset/msm-poweroff.c
+++ b/drivers/power/reset/msm-poweroff.c
@@ -288,6 +288,24 @@ static void msm_restart_prepare(const char *cmd)
 					     restart_reason);
 		} else if (!strncmp(cmd, "edl", 3)) {
 			enable_emergency_dload_mode();
+		} else if (!strncmp(cmd, "post-wdt", 8)) {
+			/* set  flag in PMIC to nofity BL post watchdog reboot */
+			qpnp_pon_store_extra_reset_info(RESET_EXTRA_POST_REBOOT_MASK,
+				RESET_EXTRA_POST_WDT_REASON);
+			 /* force cold reboot */
+			qpnp_pon_system_pwr_off(PON_POWER_OFF_HARD_RESET);
+		} else if (!strncmp(cmd, "post-pmicwdt", 12)) {
+			/* set  flag in PMIC to nofity BL post pmic watchdog reboot */
+			qpnp_pon_store_extra_reset_info(RESET_EXTRA_POST_REBOOT_MASK,
+				RESET_EXTRA_POST_PMICWDT_REASON);
+			 /* force cold reboot */
+			qpnp_pon_system_pwr_off(PON_POWER_OFF_HARD_RESET);
+		} else if (!strncmp(cmd, "post-panic", 10)) {
+			/* set  flag in PMIC to nofity BL post panic reboot */
+			qpnp_pon_store_extra_reset_info(RESET_EXTRA_POST_REBOOT_MASK,
+				RESET_EXTRA_POST_PANIC_REASON);
+			 /* force cold reboot */
+			qpnp_pon_system_pwr_off(PON_POWER_OFF_HARD_RESET);
 		} else {
 			__raw_writel(0x77665501, restart_reason);
 		}
diff --git a/include/linux/qpnp/power-on.h b/include/linux/qpnp/power-on.h
index 5c6c38d835c..6077149c87d 100644
--- a/include/linux/qpnp/power-on.h
+++ b/include/linux/qpnp/power-on.h
@@ -54,6 +54,10 @@ enum pon_restart_reason {
 	PON_RESTART_REASON_RTC		= 0x03,
 };
 
+#define RESET_EXTRA_POST_PANIC_REASON	(BIT(4) | BIT(5))
+#define RESET_EXTRA_POST_PMICWDT_REASON	BIT(5)
+#define RESET_EXTRA_POST_WDT_REASON	BIT(4)
+#define RESET_EXTRA_POST_REBOOT_MASK	(BIT(4) | BIT(5) | BIT(6))
 #define RESET_EXTRA_PANIC_REASON	BIT(3)
 #define RESET_EXTRA_REBOOT_BL_REASON	BIT(2)
 
-- 
2.11.0

